File: DFS_BFS/444-dfsForOneConnectedComponent.py

from collections import defaultdict
class Solution:
    def solve(self,matrix):
        if not matrix:
            return 0
        
        rows=len(matrix)
        data=defaultdict(int)
        cols=len(matrix[0])
        
        def dfs(r,c,x):
            if r<0 or c<0 or r>=rows or c>=cols:
                return 
            if matrix[r][c]!=x:
                return 
            matrix[r][c]=-1
            dfs(r-1,c,x)
            dfs(r+1,c,x)
            dfs(r,c-1,x)
            dfs(r,c+1,x)
            
        for r in range(rows):
            for c in range(cols):
                if matrix[r][c]!=-1:
                    color=matrix[r][c]
                    dfs(r,c,color)  #must be color here
                    data[color]+=1
        return sum(data.values())-max(data.values())
    
matrix = [
    [1, 1, 1, 1],
    [2, 2, 2, 2],
    [1, 3, 1, 2]
]


print(Solution().solve(matrix))

File: DFS_BFS/444-dfsForOneConnectedComponent.cpp
#include<iostream>
#include<vector>
#include<map>

using MATRIX=std::vector<std::vector<int>>;
using MAP=std::map<int,int>;

void f(int r,int c,int color, MATRIX& m)
{
    int R=m.size(),C=m[0].size();
    if(r<0||r>=R||c<0||c>=C)
        return ;
    if(m[r][c]!=color)
        return ;
    m[r][c]=-1;
    f(r+1,c,color,m);
    f(r-1,c,color,m);
    f(r,c+1,color,m);
    f(r,c-1,color,m);
}

int steps(MATRIX& m)
{
    MAP data{};
    int R=m.size(),C=m[0].size();
    for(int r=0;r<R;++r)
    {
        for(int c=0;c<C;++c)
        {
            if(m[r][c]!=-1)
            {
                int color=m[r][c];
                f(r,c,color,m);
                data[color]+=1;
            }
        }
    }
    int _max=0;
    int _sum=0;
    for(auto &[k,v]:data)
    {
        _sum+=v;
        if(v>_max)
            _max=v;
    }
    return _sum-_max;
}

int main()
{
    MATRIX m{
    {1, 1, 1, 1},
    {2, 2, 2, 2},
    {1, 3, 1, 2}};
    std::cout<<steps(m);
}